sources/filesystem/gitpythonfs/gitpythonfs/core.py

```diff
@@ -0,0 +1,210 @@
+                Defaults to head of the local repo.
+        """
+        super().__init__(**kwargs)
+        self.repo_path = path
+        self.repo = git.Repo(self.repo_path)
+
+    @classmethod
+    def _strip_protocol(cls, path: str) -> str:
+        path = super()._strip_protocol(path).lstrip("/")
+        if ":" in path:
+            path = path.split(":", 1)[1]
+        if "@" in path:
+            path = path.split("@", 1)[1]
+        return path.lstrip("/")
+
+    # ToDo support arguments in url, like this example from git fsspec implementation:
+    @staticmethod
+    def _get_kwargs_from_urls(path: str) -> Dict[str, str]:
+        if path.startswith("gitpythonfs://"):
+            path = path[14:]
+        out = {}
+        if ":" in path:
+            out["path"], path = path.split(":", 1)
+        if "@" in path:
+            out["ref"], path = path.split("@", 1)
+        return out
+
+    def _git_type_to_file_type(self, object: git.Object) -> str:
+        if isinstance(object, git.Blob):
+            return "file"
+        elif isinstance(object, git.Tree):
+            return "directory"
+        else:
+            msg = f"There is no fileystem object type corresponding to Git object type: {type(object).__name__}"
+            raise TypeError(msg)
+
+    def _details(
+        self, object: git.Object, include_committed_date: bool = True
+    ) -> Dict[str, Union[str, int]]:
+        """
+        Retrieves the details of a Git object.
+
+        Args:
+            object (git.Object): The Git object to retrieve details for.
+            include_committed_date (bool, optional): Whether to include the committed date. Defaults to True.
+                Getting the committed date is an expensive operation and will slow down
+                walk(), a method that is extensively used by fsspec for find(), glob() etc.
+
+        Returns:
+            dict: A dictionary containing the details typical for fsspec.
+        """
+        # commit=next(self.repo.iter_commits(paths=object.path, max_count=1))
+        details = {
+            "name": object.path,
+            "type": self._git_type_to_file_type(object),
+            "mime_type": object.mime_type if isinstance(object, git.Blob) else None,
+            "size": object.size,
+            "hexsha": object.hexsha,
+            # "committed_date": commit.committed_date,
+        }
+
+        if include_committed_date:
+            commit = next(self.repo.iter_commits(paths=object.path, max_count=1))
+            details["committed_date"] = commit.committed_date
+
+        return details
+
+    def ls(
+        self, path: str, detail: bool = False, ref: str = None, **kwargs: Any
+    ) -> Union[List[str], List[Dict]]:  # Todo implement ref
+        """List files at given path in the repo."""
+        path = self._strip_protocol(path)
+        results = []
+
+        # For traversal, always start at the root of repo.
+        tree = self.repo.tree()
```

Comment:
Done in [14eaa82](https://github.com/dlt-hub/verified-sources/pull/312/commits/14eaa8264cea1423e9aa5a86dd8f5e3742396917)  (link is not in PR context)

Since git refs feature was added the trees differ by commit, so the cache is a little more complex than setting a variable in init.  

There is some danger of the user getting unexpected results if they move a pointer (branch, HEAD) during the lifetime of an fsspec instance (noting that fsspec caches instances by defaul).   For now, I have simply mentioned that in docstrings as I believe it's an edge case.   If it's a common case, the cache could be indexed by commit sha, with conversion functions between ref <--> sha. But I'm wary of spawning more git commands just to manage the cache :) !